### ISSUE TITLE
nucleus top_p sampling seems wrong?

### ISSUE THREAD

**Issue by honglu2875** [AI-Hypercomputer/maxtext#950]
> In practice this means selecting the highest probability tokens whose cumulative probability mass
> exceeds the pre-chosen threshold p.

I think the original paper is saying that 
- we order tokens by probability from high to low
- do a cumulative sum of probability until the threshold p and stop
- accept the tokens along the process, and discard the tokens that hasn't been reached.

However, in the [function](https://github.com/AI-Hypercomputer/maxtext/blob/4aeabf193efd78af0fb8b865b0d2ac6cd6cbf8a9/MaxText/inference_utils.py#L51) that performs the nucleus top_p sampling, it does the first two steps right, but instead of discarding tokens beyond certain index in its descending order, it chooses a "cutoff logit":

https://github.com/AI-Hypercomputer/maxtext/blob/4aeabf193efd78af0fb8b865b0d2ac6cd6cbf8a9/MaxText/inference_utils.py#L63

and then it erases all the tokens whose "logits" (instead of indices in the descending ordering!) are less than this particular cutoff element

https://github.com/AI-Hypercomputer/maxtext/blob/4aeabf193efd78af0fb8b865b0d2ac6cd6cbf8a9/MaxText/inference_utils.py#L64

I don't see how this can be correct, unless I misunderstood the top_p sampling algorithm?

**Comment by honglu2875** [comment]
ahh I was being stupid. Removing and closing as it was fully due to my carelessness.
